### ISSUE TITLE
Arbitrary currency argument

### ISSUE THREAD

**Issue by dvtomas** [format-message/format-message#113]
Hi,
maybe I've just missed it in the docs. Is it possible to insert arbitrary currency as parameter? Something along the lines

```javascript
formatMessage(
  'This Hello to you costs {price, number, currency} Czech korunas!', 
  { price: {'type': 'currencyAmount', 'code': 'CZK', 'amount': 1.0 }}
)
``` 
=> 
```"This Hello to you costs 1.0 CZK"``` 
(or perhaps
```"This Hello to you costs 1,0 Kč"``` 
depending on locale)

icu4java has a mechanism for it: http://icu-project.org/apiref/icu4j/com/ibm/icu/util/CurrencyAmount.html

Thanks, cheers

**Comment by thetalecrafter** [comment]
If you know the currency you will use at build time:

```
formatMessage.setup({ formats: { number: {
  korunas: {
    currency: 'CZK'
  }
} } })

formatMessage('This Hello to you costs {price, number, korunas}!', { price: 1.0 })
```

If you need to determine the currency at run time, you could use a generic string placeholder, and format it with the Intl.NumberFormat or .toLocaleString:

```
formatMessage('This Hello to you costs {price}!', {
  price: priceNum.toLocaleString(lang, { currency: currentCurrency })
})
```

I should probably add these examples to some documentation, since it's not that obvious.

**Comment by dvtomas** [comment]
The currency will be determined at runtime. I guess the second approach will do for now, I'll reopen the issue if I bump into problems. Thanks!

**Comment by dvtomas** [comment]
Sorry, I've accidentally closed the issue, reopening so you can update the docs and close it yourself :(

**Comment by dvtomas** [comment]
Just by the way - there's a use case I personally will probably not need, but first class currency support à la http://icu-project.org/apiref/icu4j/com/ibm/icu/util/CurrencyAmount.html could be useful for.

There is a web backend and frontend. The backend communicates message keys and expansion arguments to the frontend. Frontend looks up the key in it's own message catalog and expands the localized message with the arguments. I can easily see the need for currency amount (number + currency) arguments in the message - imagine a localizable site displaying exchange rates for various currencies. An example of a hypothetical message traveling from backend to frontend:

```javascript
{
  messageKey: "1 US dollar = {rate, number, currency}", 
  args: {price: {type: "currency", currency: "CZK", value: 24.8}
}
```

which the frontend would use as

```javascript
formatMessage(translate(messageKey), args)
```

**Comment by thetalecrafter** [comment]
The `next` branch has support for number patterns, and these can include currency symbols. In fact `{ price, number, CZK }` should work. That would still be a currency known at build-time, rather than run-time, though.

It also supports custom placeholder types, so you could define a `currency` type:
```js
formatMessage.setup({
  types: {
    currency: (placeholder, locales) => (arg, args) =>
      Number(arg.value).toLocaleString(locales, {
        style: 'currency',
        currency: arg.currency
      })
  }
})
formatMessage('1 US dollar = {rate, currency}', { rate: { value: 24.8, currency: 'CZK' } })
````

Do either of those look like they'd work well for you?
I haven't released these changes yet, so now would be a great time to provide feedback.

**Comment by dvtomas** [comment]
Hey,
sorry, I've changed my job couple of months ago and I'm doing something very different now. I'm afraid I can't provide you with any feedback. I value your effort, though, thank you.

**Comment by thetalecrafter** [comment]
currency formatting available in v6.0.0
